### PR TITLE
Run all proof harnesses by default

### LINKED ISSUE #464: Add a `#[proof]` attribute that allow us to mark proof harnesses
**Requested feature:** Create an attribute that allow users to mark their proof harness functions.
**Use case:** The same way users can tag their test functions using the `#[test]` attribute, users should be able to tag their proof harness with an attribute such as `#[proof]`. We can then create a cargo subcommand, like `cargo verify` that will execute all proofs in a crate.

**Link to relevant documentation (Rust reference, Nomicon, RFC)**: N/A
**Is this a breaking change?** No

**Test case**:

We could modify the test in src/test/expected/enum/main.rs to be something like:

```rust
// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0 OR MIT

/* other stuff */

#[cfg(rmc)]
mod rmc_tests {
    use super::*;
    #[proof]
    fn test_one_plus_two() {
        let p = Pair::new(1, 2);
        assert!(p.sum() == 3);
    }
}
```

### LINKED ISSUE #691: End-to-end tests needed for proof harness attributes
(Partially a note to self here)

We have a test in `src/test/cargo-rmc/simple-proof-annotation` that runs the code for proof harness attributes, but we don't have any tests checking that it produces the correct answer.

This test suite has special merit for ensuring that we're actually recognizing proof annotations is all cases. We had an issue early on where the function weren't being codegen'd at all, and so we never saw them to take note of the proof annotation.

So we should include: public functions, private functions, crate-type bin, crate-type lib, conflicting names, and whatever else we think of.

### PR DESCRIPTION
**Pull request by tedinski**
### Description of changes: 

1. Hide (but technically keep) `--function` in favor of `--harness` which must be `#[kani::proof]`. Updated all documentation accordingly.
2. cargo-kani now parses and merges `kani-metadata.json` files, to support looking up proof harnesses.
    * Quick note: I changed dry-run output to report `*.symtab.json` instead of `dry-run.symtab.json`
4. (Minor: I uh, altered `--gen-c` to print where the file was generated.)
5. `crate-type=lib` is now unconditional and we (by documentation) require `#[kani::proof]`. Technically, `--function` can still specify anything but you need some method of ensuring the function gets code-gen'd.
6. THE MEAT: **`cargo-kani` and `kani` now default to running all proof harnesses found.**
    * Switch to just one with `--harness`
    * Behavior is to run each one one after another until either one fails (and we stop) or all suceed and we print "All harnesses successfully verified."
7. Added a new mode for `cargo-kani` mode tests. We can now add a plain `expected` file instead of `function.expected` and this will run `cargo-kani` without any `--function` flags.

phew

### Resolved issues:

Resolves #464
Resolves #691 

Contributes to #477 and #701

### Testing:

* How is this change tested? 2 new tests cases, fairly basic. Plus existing whole test suite.

* Is this a refactor change? no

### Checklist
- [x] Each commit message has a non-empty body, explaining why the change was made
- [x] Methods or procedures are documented
- [x] Regression or unit tests are included, or existing tests cover the modified code
- [x] My PR is restricted to a single feature or bugfix

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 and MIT licenses.

### CODE REVIEW

**Comment by tedinski** [comment]
oh, doh. I need to adapt bookrunner to work with needing `#[kani::proof]` annotations now. Welp, monday.

File: src/cargo-kani/src/call_goto_instrument.rs [hunk 1]
```diff
@@ -28,6 +24,7 @@ impl KaniSession {
         }
 
         if self.args.gen_c {
+            println!("Wrote {}", output.to_string_lossy());
```
Comment (celinval) [inline src/cargo-kani/src/call_goto_instrument.rs:27]: Is this meant to be user visible? If so, can we expand the message a bit more? Also, should we check if we're in quiet mode?

File: src/cargo-kani/src/call_goto_instrument.rs [hunk 2]
```diff
@@ -124,77 +121,3 @@ impl KaniSession {
         self.run_suppress(cmd)
     }
 }
-
-/// Collect all vtable restriction metadata together, and write one combined output in CBMC's format
-fn link_function_pointer_restrictions(
```
Comment (celinval) [inline src/cargo-kani/src/call_goto_instrument.rs:129]: Next time, I would prefer if you leave the refactoring of unrelated code to separate PR. Thanks

File: src/cargo-kani/src/main.rs [hunk 3]
```diff
@@ -49,20 +51,34 @@ fn cargokani_main(mut input_args: Vec<OsString>) -> Result<()> {
...
+        }
+
+        if ctx.args.visualize {
+            ctx.run_visualize(&specialized_obj, "target/report")?;
+        } else {
+            let result = ctx.run_cbmc(&specialized_obj)?;
+            if result == VerificationStatus::Failure {
+                // Failure exit code without additional error message
+                drop(ctx);
+                std::process::exit(1);
```
Comment (celinval) [inline src/cargo-kani/src/main.rs:72]: Can we run all harnesses first instead of stopping after the first failure?

File: src/cargo-kani/src/main.rs [hunk 4]
```diff
@@ -83,17 +99,37 @@ fn standalone_main() -> Result<()> {
...
-    if ctx.args.visualize {
-        ctx.run_visualize(&linked_obj, "report")?;
-    } else {
-        let result = ctx.run_cbmc(&linked_obj)?;
-        if result == VerificationStatus::Failure {
-            // Failure exit code without additional error message
-            drop(ctx);
-            std::process::exit(1);
+
+    let metadata = ctx.collect_kani_metadata(&[outputs.metadata])?;
```
Comment (celinval) [inline src/cargo-kani/src/main.rs:103]: Can we move the duplicated code to a separate function?

File: src/cargo-kani/src/metadata.rs [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+                return Ok(vec![harness.clone()]);
+            }
+            bail!("A proof harness named {} was not found", name);
+        }
+        if metadata.proof_harnesses.is_empty() {
+            // This is not the desirable behavior we deliver to our users,
+            // but this is a hack to get kani regression tests back to working quickly.
+            // TODO: This should be removed in favor of a good error message suggesting that proof harnesses be written.
+            // Probably even with a link to Kani documentation!
+            Ok(vec![mock_proof_harness("main")])
```
Comment (celinval) [inline src/cargo-kani/src/metadata.rs:141]: Is this still needed? I thought you had added annotations to all tests already.

File: tests/expected/multiple-harnesses/expected [hunk 6]
```diff
@@ -0,0 +1,3 @@
+Checking harness check_first_harness...
```
Comment (celinval) [inline tests/expected/multiple-harnesses/expected:1]: Can you please move these tests to be under `ui` suite?

File: tools/compiletest/src/runtest.rs [hunk 7]
```diff
@@ -317,10 +317,12 @@ impl<'test> TestCx<'test> {
         let function_name = self.testpaths.file.file_stem().unwrap().to_str().unwrap();
         cargo
             .arg("kani")
-            .args(["--function", function_name])
             .arg("--target-dir")
             .arg(self.output_base_dir().join("target"))
             .current_dir(&parent_dir);
+        if "expected" != self.testpaths.file.file_name().unwrap() {
+            cargo.args(["--function", function_name]);
```
Comment (celinval) [inline tools/compiletest/src/runtest.rs:324]: Do we have a jira to change these to `--harness`?

**Review by celinval — CHANGES_REQUESTED** [review]
Sweet! Mostly minor comments. I do think we should change the behaviour to run all harnesses before failing though.

File: src/cargo-kani/src/args.rs [hunk 8]
```diff
@@ -72,9 +72,12 @@ pub struct KaniArgs {
...
     pub checks: CheckArgs,
 
-    /// Entry point for verification
-    #[structopt(long, default_value = "main")]
-    pub function: String,
+    /// Entry point for verification (symbol name)
+    #[structopt(long, hidden = true)]
+    pub function: Option<String>,
+    /// Entry point for verification (proof harness)
+    #[structopt(long, conflicts_with = "function", conflicts_with = "dry-run")]
```
[the remark this answers is not in the document]
Comment (celinval) [reply src/cargo-kani/src/args.rs:80]: But do we need to? We can just print something like dummy in the function name, right?

I wonder if we should make the dry-run a hidden option though.

File: tools/compiletest/src/runtest.rs [hunk 9]
```diff
@@ -280,15 +280,17 @@ impl<'test> TestCx<'test> {
...
+        match self.props.kani_panic_step {
+            Some(KaniFailStep::Check) => {
+                self.check();
+            }
+            Some(KaniFailStep::Codegen) => {
+                self.codegen();
+            }
+            Some(KaniFailStep::Verify) | None => {
+                self.verify();
+            }
```
Comment (tedinski) [inline tools/compiletest/src/runtest.rs:292]: This change is necessary since we now may run `kani-compiler` from `kani` with different flags that `compiletest` wasn't picking up. This change means we avoid doing redundant work, and these differences only matter for things that expected to fail to check or codegen (which wasn't any existing tests, thankfully)

Just thought I'd note this in the review here, since it got mixed up with a bunch of other changes to get tests working again after switching to using `--harness`

File: tools/compiletest/src/runtest.rs (hunk 7 above)
Comment (tedinski) [reply tools/compiletest/src/runtest.rs]: Changed them to harness now.

File: src/cargo-kani/src/main.rs (hunk 4 above)
Comment (tedinski) [reply src/cargo-kani/src/main.rs:103]: I've tried to avoid having any functions that are specific to `kani` or `cargo-kani` (except these two of course) because I think these are generally design errors. You'll notice that there are several small variations between these two functions, and trying to abstract those differences out will generally make the code worse, not better.

A small amount of not-really-duplication is better than trying to create a function with a half dozen parameters that only gets called 2 ways. The old python was full of that, and it's just code that only gets worse as people add to it.

That said, I think there IS a little bit I can factor out here, I just need to fix how we do report generation first since I actually broke that...

Comment (celinval) [reply src/cargo-kani/src/main.rs:103]: That makes sense. Thanks

**Comment by tedinski** [comment]
This should be ready to review again.

It now runs all proof harnesses even if one fails, and prints a summary at the end:

```
Complete - 2 successfully verified harnesses, 0 failures, 2 total.
```

Hmm. Possibly I should print which harnesses failed, so you don't have to scroll up to search for them... Let me know your thoughts.

We now also produce all reports with `--visualize` which required creating multiple directories, one for each harness:

```
$ cargo kani --visualize
$ ls -l target/
total 24
-rw-rw-r-- 1 ubuntu ubuntu  177 Mar 23 19:13 CACHEDIR.TAG
drwxrwxr-x 7 ubuntu ubuntu 4096 Mar 23 19:13 debug
drwxrwxr-x 4 ubuntu ubuntu 4096 Mar 23 19:22 report-proofs::check_trivial
drwxrwxr-x 4 ubuntu ubuntu 4096 Mar 23 19:22 report-proofs::failed_check
drwxrwxr-x 4 ubuntu ubuntu 4096 Mar 23 19:22 report-proofs::third_harness
drwxrwxr-x 3 ubuntu ubuntu 4096 Mar 23 19:13 x86_64-unknown-linux-gnu
```

I notice that the "pretty name" of function is fully qualified. I'm thinking about fixing that (advice?) but if we don't use the fully qualified names, there might be name collisions. (At the very least, colon isn't legal in filenames on windows, so we have to do something...)

**Comment by celinval** [comment]
The way proof is implemented today by using `no_mangle` guarantees that you can't have name collision between harnesses. I believe if that happens, the compilation fails.

That said, can you replace the `::` by `.`?

File: src/cargo-kani/src/metadata.rs [hunk 10]
```diff
@@ -0,0 +1,193 @@
...
+    name: &str,
+    harnesses: &'a [HarnessMetadata],
+) -> Option<&'a HarnessMetadata> {
+    for h in harnesses.iter() {
+        if h.pretty_name == *name {
+            return Some(h);
+        } else {
+            // pretty_name will be things like `module::submodule::name_of_function`
+            // and we want people to be able to specify `--harness name_of_function`
+            if let Some(prefix) = h.pretty_name.strip_suffix(name) {
```
Comment (zhassan-aws) [inline src/cargo-kani/src/metadata.rs:165]: We should probably check for the case where there is ambiguity, i.e., multiple functions that have the same name, e.g. `module1::submodule1::name_of_function` and `module2::submodule2::name_of_function`, and emit an error.

File: tests/smack/basic/arith.rs [hunk 11]
```diff
@@ -1,8 +1,8 @@
 // Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 // SPDX-License-Identifier: Apache-2.0 OR MIT
-// @expect verified
```
Comment (zhassan-aws) [inline tests/smack/basic/arith.rs:3]: Is removing the SMACK annotations necessary? They're probably meant for the SMACK testing tool, but they may be useful for us to compare results.

File: tools/bookrunner/src/util.rs [hunk 12]
```diff
@@ -203,7 +203,8 @@ pub fn add_codegen_job(litani: &mut Litani, test_props: &TestProps) {
 pub fn add_verification_job(litani: &mut Litani, test_props: &TestProps) {
     let exit_status = if test_props.fail_step == Some(FailStep::Verification) { 10 } else { 0 };
     let mut kani = Command::new("kani");
-    kani.arg(&test_props.path).args(&test_props.kani_args);
+    // Some of test_props.kani_args will contains `--cbmc-args` so we should always put that last
+    kani.arg(&test_props.path).args(&["--function", "main"]).args(&test_props.kani_args);
```
Comment (zhassan-aws) [inline tools/bookrunner/src/util.rs:207]: Are we adding `--function main` to avoid having to annotate the bookrunner examples with `#[kani::proof]`? If so, can we add a comment?

**Review by zhassan-aws — APPROVED** [review]
This is a huge step forward! Awesome work, @tedinski!

Just had a few minor comments.

File: src/cargo-kani/src/main.rs [hunk 13]
```diff
@@ -83,20 +94,81 @@ fn standalone_main() -> Result<()> {
...
+    ) -> Result<()> {
+        if !self.args.quiet && !self.args.visualize && harnesses.len() > 1 {
+            if !failed_harnesses.is_empty() {
+                println!("Summary:");
+            }
+            for harness in failed_harnesses.iter() {
+                println!("Verification failed for - {}", harness.pretty_name);
+            }
+
+            println!(
```
[the remark this answers is not in the document]
Comment (tedinski) [reply src/cargo-kani/src/main.rs:156]: Maybe. In my experience, there is a contingent of "if I say quiet, actually print nothing, just give me a return value" software out there. But... maybe this isn't one that should be that way.

cargo test is a compelling point of comparison. Possibly we should do some output comparisons and decide differently later?

File: tests/smack/basic/arith.rs (hunk 11 above)
Comment (tedinski) [reply tests/smack/basic/arith.rs:3]: No: I needed to quickly audit those files to fix a few things and I removed what I thought was unnecessary. I don't see how these would ever get used...

File: src/cargo-kani/src/main.rs (hunk 13 above)
[the remark this answers is not in the document]
Comment (zhassan-aws) [reply src/cargo-kani/src/main.rs:156]: Sure.
